### ISSUE TITLE
Implement `Grid.build_node_face_connectivity`

### ISSUE THREAD

**Issue by anissa111** [UXARRAY/uxarray#130]
Implement `Grid.build_node_face_connectivity`

**Comment by anissa111** [comment]
@paullric I'm working on this implementation now, could you provide a little clarification about what this function is supposed to do?

Does the function build `node_face_connectivity` from scratch? Or is this function just building a reverse relationship from a pre-existing `face_node_connectivity`?

**Comment by paullric** [comment]
It should have to build it from scratch, if face_node_connectivity is unavailable.  If it is available then you can leverage it and transpose.

~ Paul

**Comment by anissa111** [comment]
> It should have to build it from scratch, if face_node_connectivity is unavailable. If it is available then you can leverage it and transpose.

Oh, okay, I see thank you! Is there a prefered method for doing this calculation or is something similar implemented in `tempestremap` I should look at?

**Comment by Huite** [comment]
Hi @anissa111,

I butt in here occasionally, I've implemented this in `xugrid` if you're looking for inspiration:

https://github.com/Deltares/xugrid/blob/main/xugrid/ugrid/ugrid2d.py#L406

It mostly uses some utils defined here:
https://github.com/Deltares/xugrid/blob/main/xugrid/connectivity.py

Which I find make it relatively easy to these transformations on connectivity arrays.

In general, I found it convenient to use scipy.sparse functionality to build/represent these arrays here and there. In particular, constructing a CSR (https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html) matrix does a lot of sorting (which is C++ behind the scenes). Alternatively, you require a lot of numpy sorting calls, which aren't that easy to do efficiently if the arrays are really not that rectangular; or you'd have to write a lot of numba code, I suppose.
